qi-lib/flow/core/compiler/deforest/syntax.rkt

```diff
@@ -0,0 +1,298 @@
+
+(define-syntax-class fsc-empty?
+  #:literal-sets (fs-literals)
+  #:literals (qi:null? qi:empty?)
+  (pattern (esc
+            (#%host-expression (~or qi:empty?
+                                    qi:null?))))
+  (pattern (#%fine-template
+            ((#%host-expression (~or qi:empty?
+                                     qi:null?)) _)))
+  (pattern (#%blanket-template
+            ((#%host-expression (~or qi:empty?
+                                     qi:null?)) __))))
+
+(define-syntax-class fsc-default
+  #:datum-literals (cstream->list)
+  (pattern cstream->list))
+
+(define-syntax-class fsc-syntax
+  (pattern (~or _:fsc-foldr
+                _:fsc-foldl
+                _:fsc-list-ref
+                _:fsc-length
+                _:fsc-empty?
+                _:fsc-default
+                )))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; The actual fusion generator implementation
+
+;; Used only in deforest-rewrite to properly recognize the end of
+;; fusable sequence.
+(define-syntax-class non-fusable
+  (pattern (~not (~or _:fst-syntax
+                      _:fsp-syntax
+                      _:fsc-syntax))))
+
+(define (make-deforest-rewrite generate-fused-operation)
+  (lambda (stx)
+    (syntax-parse stx
+      [((~datum thread) _0:non-fusable ...
+                        p:fsp-syntax
+                        ;; There can be zero transformers here:
+                        t:fst-syntax ...
+                        c:fsc-syntax
+                        _1 ...)
+       #:with fused (generate-fused-operation
+                     (syntax->list #'(p t ... c))
+                     stx)
+       #'(thread _0 ... fused _1 ...)]
+      [((~datum thread) _0:non-fusable ...
+                        t1:fst-intf0
+                        t:fst-syntax ...
+                        c:fsc-syntax
+                        _1 ...)
+       #:with fused (generate-fused-operation
+                     (syntax->list #'(list->cstream t1 t ... c))
+                     stx)
+       #'(thread _0 ... fused _1 ...)]
+      [((~datum thread) _0:non-fusable ...
+                        p:fsp-syntax
+                        ;; Must be 1 or more transformers here:
+                        t:fst-syntax ...+
+                        _1 ...)
+       #:with fused (generate-fused-operation
+                     (syntax->list #'(p t ... cstream->list))
+                     stx)
+       #'(thread _0 ... fused _1 ...)]
+      [((~datum thread) _0:non-fusable ...
+                        f1:fst-intf0
+                        f:fst-syntax ...+
+                        _1 ...)
+       #:with fused (generate-fused-operation
+                     (syntax->list #'(list->cstream f1 f ... cstream->list))
+                     stx)
+       #'(thread _0 ... fused _1 ...)]
+      ;; return the input syntax unchanged if no rules
+      ;; are applicable
+      [_ stx])))
+
+(define-syntax (define-and-register-deforest-pass stx)
```

Comment:
I'm not convinced this macro is necessary as it's used in only one place.

Reply:
My plan is to use this macro in both CPS and named-let runtime implementation. Therefore I propose we keep it and move it to the deforest/fusion.rkt module like the make-deforest-rewrite syntax.